addons/sale/views/sale_order_views.xml

```diff
@@ -377,13 +381,26 @@
                                 <group>
                                     <group invisible="display_type">
                                         <field name="product_updatable" invisible="1"/>
-                                        <field name="product_id"
-                                            domain="[('sale_ok', '=', True)]"
-                                            context="{'partner_id':parent.partner_id, 'quantity':product_uom_qty, 'pricelist':parent.pricelist_id, 'uom':product_uom, 'company_id': parent.company_id}"
-                                            readonly="not product_updatable"
-                                            required="not display_type and not is_downpayment"
-                                            force_save="1"
-                                            widget="many2one_barcode"/>
+                                        <div class="d-flex align-items-baseline">
+                                            <span class="fa fa-exclamation-triangle text-warning me-1"
+                                                title="This line contains archived product"
+                                                invisible="not has_archived_product"
+                                            />
+                                            <field name="product_id"
+                                                domain="[('sale_ok', '=', True)]"
+                                                context="{
+                                                    'partner_id':parent.partner_id,
+                                                    'quantity':product_uom_qty,
+                                                    'pricelist':parent.pricelist_id,
+                                                    'uom':product_uom,
+                                                    'company_id': parent.company_id
```

Comment:
Please add a space after each `:` since you're already modifying those lines ;).